zkp/circuits/anon_enc_nullifier_non_repudiation.circom

```diff
@@ -0,0 +1,198 @@
+  // the sender really owns the private key for the input
+  // UTXOs
+  var inputOwnerPublicKey[2];
+  component pub = BabyPbk();
+  pub.in <== inputOwnerPrivateKey;
+  inputOwnerPublicKey[0] = pub.Ax;
+  inputOwnerPublicKey[1] = pub.Ay;
+  var inputOwnerPublicKeys[nInputs][2];
+  for (var i = 0; i < nInputs; i++) {
+    inputOwnerPublicKeys[i][0] = inputOwnerPublicKey[0];
+    inputOwnerPublicKeys[i][1] = inputOwnerPublicKey[1];
+  }
+
+  component checkPositives = CheckPositive(nOutputs);
+  checkPositives.outputValues <== outputValues;
+
+  component checkInputHashes = CheckHashes(nInputs);
+  checkInputHashes.commitments <== inputCommitments;
+  checkInputHashes.values <== inputValues;
+  checkInputHashes.salts <== inputSalts;
+  checkInputHashes.ownerPublicKeys <== inputOwnerPublicKeys;
+
+  component checkOutputHashes = CheckHashes(nOutputs);
+  checkOutputHashes.commitments <== outputCommitments;
+  checkOutputHashes.values <== outputValues;
+  checkOutputHashes.salts <== outputSalts;
+  checkOutputHashes.ownerPublicKeys <== outputOwnerPublicKeys;
+
+  component checkNullifiers = CheckNullifiers(nInputs);
+  checkNullifiers.nullifiers <== nullifiers;
+  checkNullifiers.values <== inputValues;
+  checkNullifiers.salts <== inputSalts;
+  checkNullifiers.ownerPrivateKey <== inputOwnerPrivateKey;
+
+  component checkSum = CheckSum(nInputs, nOutputs);
+  checkSum.inputValues <== inputValues;
+  checkSum.outputValues <== outputValues;
+
+  // With the above steps, we demonstrated that the nullifiers
+  // are securely bound to the input commitments. Now we need to
+  // demonstrate that the input commitments belong to the Sparse
+  // Merkle Tree with the root `root`.
+  component checkSMTProof = CheckSMTProof(nInputs, nSMTLevels);
+  checkSMTProof.root <== root;
+  checkSMTProof.merkleProof <== merkleProof;
+  checkSMTProof.enabled <== enabled;
+  checkSMTProof.leafNodeIndexes <== inputCommitments;
+
+  // generate shared secret for the receiver
+  var sharedSecretReceiver[2];
+  component ecdh1 = Ecdh();
+  ecdh1.privKey <== inputOwnerPrivateKey;
+  // our circuit requires that the output UTXO for the receiver must be the first in the array
+  ecdh1.pubKey[0] <== outputOwnerPublicKeys[0][0];
+  ecdh1.pubKey[1] <== outputOwnerPublicKeys[0][1];
+  sharedSecretReceiver[0] = ecdh1.sharedKey[0];
+  sharedSecretReceiver[1] = ecdh1.sharedKey[1];
+
+  // encrypt the value for the receiver
+  component encrypt1 = SymmetricEncrypt(2);
+  // our circuit requires that the output UTXO for the receiver must be the first in the array
+  encrypt1.plainText[0] <== outputValues[0];
+  encrypt1.plainText[1] <== outputSalts[0];
+  encrypt1.key <== sharedSecretReceiver;
+  encrypt1.nonce <== encryptionNonce;
+  encrypt1.cipherText[0] --> cipherText[0];
+  encrypt1.cipherText[1] --> cipherText[1];
+
+  // generate shared secret for the authority
+  var sharedSecretAuthority[2];
+  component ecdh2 = Ecdh();
+  ecdh2.privKey <== inputOwnerPrivateKey;
```

Comment:
I'm not grasping this, we generate a key using the inputs private key to then sign the values passed in so the in future we can check with the public key that the entity that had the private key was the one that tried to perform this transaction?

Reply:
my understanding is it's used by ECDH to compute the shared key

Reply:
But we also have an input into the circuit of the authority public key

Reply:
It's a key pair of the public key provided and the private key of the input owner?

Reply:
the shared secret is used to encrypt rather than sign, the goal is so that the receiver of the encrypted values (the one whose public key is used here) can use their private key and the sender's public key, to derive the same shared secret (this is per the Diffie-Hellman key agreement protocol) to decrypt

Reply:
Gotta now!

- The inputs are signed by a shared key of the initiator so their private key and the public key of the regulator so only the regulator can in the future decipher them
- The regulator can then verify that the cipher matches all the other inputs into the circuit by decrypting with a shared key of the regulator private key and the initiator public key!

Awesome

Reply:
> we also have an input into the circuit of the authority public key
yes that's the right observation. we are performing two separate encryptions, targeted for two identities (the receiver of the output UTXO, as well as the authority). so we use two separate ECDH to deriver two separate shared secrets to do the encryption